FR #1090: Associate attributes with datasets (#1431)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: specification/datasets/cost_and_usage/dataset.md

| [SKU Price ID](#skupriceid)                                                   | Dimension   | Conditional   | True         | String    |
| [Sub Account ID](#subaccountid)                                               | Dimension   | Conditional   | True         | String    |
| [Sub Account Name](#subaccountname)                                           | Dimension   | Conditional   | True         | String    |
| [Sub Account Type](#subaccounttype)                                           | Dimension   | Conditional   | True         | String    |
| [Tags](#tags)                                                                 | Dimension   | Conditional   | True         | JSON      |

<div class='h4-nonindex'>Relationships</div>

None

<div class='h4-nonindex'>Requirements</div>

The CostAndUsage dataset adheres to the following requirements:

* CostAndUsage MUST conform to [ColumnHandling](#columnhandling) requirements.
* CostAndUsage MUST conform to [DiscountHandling](#discounthandling) requirements.
* CostAndUsage MUST conform to [NullHandling](#nullhandling) requirements.

> 

Author: Per [this thread](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/1321#discussion_r2337300515) on PR #1321, we will add the following line if/when #1321 is approved:

```suggestion
CostAndUsage MUST conform to [InvoiceHandling](#invoicehandling) requirements.

```

Author: This line will be added to #1321, now that this PR has been approved.  Resolving.

> <div class='h4-nonindex'>Dataset ID</div>

Reviewer: I ran into a problem in the recency PR that I want to bring up here @shawnalpay. I am going to author a change there but here is the key points of the problem that we likely need to coordinate:

- In metadata, we refer to datasets by a single key called "dataset" and i examples that is a guid
- - Because Data generators may have multiple reports using the same dataset we likely need a id reference to the FOCUS dataset, and the report that uses a focus dataset
- -  noticed that we call it Dataset ID here....which was my initial thought in the metadata section. What do you think about dataset_instance_id
- We should include a name in the metdata which is the name of the dataset in FOCUS

Author: Discussed with @rileyjenk.  He will iterate on PR #1373, and we can resolve this.

CostAndUsage

<div class='h4-nonindex'>Display Name</div>

Cost and Usage

<div class='h4-nonindex'>Description</div>

> Describes the cost and usage incurred through using or purchasing a provider’s [*resources*](#glossary:resource) or [*services*](#glossary:service).

Author: Striking smart character to fix the linter.

```suggestion
Describes the cost and usage incurred through using or purchasing a provider's [*resources*](#glossary:resource) or [*services*](#glossary:service).
```

# Cost and Usage

> The Cost and Usage dataset is the primary dataset for FOCUS cost and usage data.

Reviewer 2: I suggest we consider expanding the glossary with additional dataset-related terms, either as part of this PR or during the Consistency review phase. 

Currently, we only have a definition for a FOCUS dataset, and in all columns in the Cost and Usage dataset, presence-related normative requirements link to this single glossary term.

Patterns for column presence normative requirements:

```markdown
* <ColumnId> MUST be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset).
* <CoumnId> MUST be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset) when <Condition>.
* <ColumnId> is RECOMMENDED to be present in a [*FOCUS dataset*](#glossary:FOCUS-dataset) when <Condition>.
```

Current FOCUS dataset glossary term definition:

> *A FOCUS dataset is a structured collection of records that conforms to the BCP14 criteria established by FOCUS. All columns included must be defined in the FOCUS Columns section of the specification.*
> 
> *In addition to these standardized columns, data generators may include custom provider-specific columns (prefixed with `x_`) where additional context is needed beyond what is captured in the defined FOCUS columns. If such custom columns introduce record-splitting (i.e., a single original charge results in multiple rows), the data generator is responsible for ensuring that all cost and quantity metrics still meet the aggregation and consistency rules required by the specification.*
> 
> *The collection of datasets are designed to provide billing insight, additional context, metadata, mapping, or enrichment information that enhances the interpretability or completeness.*

Author: @ijurica What are some examples of terms you would want to add?

Reviewer 2: ***Note:***
*No need to rush this! I just noticed that we could be a bit more precise in these normative requirements*

---

#### Option A:

Reference the specific dataset in presence-related normative requirements, rather than the existing FOCUS dataset glossary term.

#### Option B:

Introduce glossary terms for individual datasets and reference them in presence-related normative requirements, instead of the existing FOCUS dataset glossary term.

Examples of glossary terms we could add:

- Cost and Usage dataset
- Contract Commitment dataset (scope of PR #1350)
.

Author: When we approved #1337, we agreed not to have separate glossary entries for each dataset (@mike-finopsorg, please chime in if you remember it differently).

I think Option A is the safer and less duplicative play.  I propose that we tackle that as part of work in FR #1051 in the next release.

Let me know if this works for you, and I will resolve this comment.

Reviewer 2: It was just a note... 
resolving this conversation.

| [SKU ID](#skuid)                                                              | Dimension   | Conditional   | True         | String    |
| [SKU Meter](#skumeter)                                                        | Dimension   | Conditional   | True         | String    |
| [SKU Price Details](#skupricedetails)                                         | Dimension   | Conditional   | True         | JSON      |
| [SKU Price ID](#skupriceid)                                                   | Dimension   | Conditional   | True         | String    |
| [Sub Account ID](#subaccountid)                                               | Dimension   | Conditional   | True         | String    |
| [Sub Account Name](#subaccountname)                                           | Dimension   | Conditional   | True         | String    |
| [Sub Account Type](#subaccounttype)                                           | Dimension   | Conditional   | True         | String    |
| [Tags](#tags)                                                                 | Dimension   | Conditional   | True         | JSON      |

> <div class='h4-nonindex'>Relationships</div>

Author: These `<div>` headers aren't rendering the way we set it up, @mike-finopsorg. 

See this PDF output (go to Section 3.1): [spec.pdf](https://github.com/user-attachments/files/22127586/spec.pdf) (screenshot below also)

Any clue why this might be?

<img width="653" height="485" alt="Image" src="https://github.com/user-attachments/assets/6aed34ed-62cc-49df-820e-a04eefbb81f2" />

Author: Discussed with @mike-finopsorg.  This doesn't work because this CSS was implemented via PR #1350 and hasn't yet hit `working_draft` 😅   My mistake!  

Working as intended, and the use of these divs will work fine once #1350 is approved.  Resolving!

> (none)

Reviewer 2: nit: Here it says **`(none)`**, while for Exceptions in Attribute definitions we used **`None`** - should we align?

Author: Happy to stay consistent!

```suggestion
None
```

<div class='h4-nonindex'>Requirements</div>

The CostAndUsage dataset adheres to the following requirements:

> * CostAndUsage MUST be present.

Reviewer 3: This doesn't seem to be needed, right? I'm not sure what it means. If they have the dataset, these rules apply. If they don't have the dataset, then these rules don't apply because they're not even looking at cost and usage data. I'm not sure what this is really conveying as a requirement.

Unless it's saying that the CostAndUsage dataset is required for them to be able to claim FOCUS support? But if that's it, what are they "certifying" is FOCUS-compatible? A dataset or a tool? I believe we can only certify datasets, so I'm not sure this line is needed.

Author: @flanakin My point here was, a Cost and Usage dataset is required in all circumstances.  Some of the supporting datasets may be required conditionally; for example, generating the Contract Commitment dataset as described in #1350 is only applicable if the data generator supports contract commitments.  

I suspect what you're getting at is: this isn't a _dataset_ requirement, but more of a _global_ requirement.  Kind of like how, per our conformance requirements work via #1209, the presence of a column is a dataset requirement, and not a column requirement.  _That_, I could agree with.  And maybe we do need a global requirement space.  But I'd probably put that out of scope for the time being and just strike this line for now, if so.

Let me know what you think about that.

Reviewer 3: If we have this, we're saying that tools are FOCUS compliant or not. And that compliance defines what datasets you MUST and SHOULD provide. 

I was assuming tools would have the ability to say they have a "FOCUS compliant cost and usage" -OR- "FOCUS compliant contract commitments" -OR- "FOCUS compliant SKU catalog", etc. 

Was there a decision about the level at which we are measuring compliance and enforcing requirements? I'm fine either way, I suspect I just came at it with a different assumption. I see pros/cons with both approaches.

Reviewer 4: This line surfaces the next entity which we discovered in the CR work. What you are alluding to here is that there is a "Data Generator" entity and it has requirements of "CostAndUsage MUST be provided by the Data Generator". I would like the group to consider skipping this normative text for this release and to put a FR in to scope requirements of the Data Generator entity in order to avoid the same issue we have today with Column requirements that should be relocated to the Dataset entity.

Author: Great points, both -- thank you @flanakin and @mike-finopsorg.  I'll strike it!  I agree that we should add an FR to scope normative requirements at the "global" (i.e., data generator) level; I'll coordinate with @Matt-Cowsert.

```suggestion
```

* CostAndUsage MUST conform to [ColumnHandling](#columnhandling) requirements.
* CostAndUsage MUST conform to [DiscountHandling](#discounthandling) requirements.
* CostAndUsage MUST conform to [NullHandling](#nullhandling) requirements.

<div class='h4-nonindex'>Dataset ID</div>

CostAndUsage

<div class='h4-nonindex'>Display Name</div>

Cost and Usage

<div class='h4-nonindex'>Description</div>

> Describes the cost and usage incurred by consuming a provider's resources and services.

Reviewer 2: I'm not sure if this wording sufficiently covers purchase scenarios.

Author: @ijurica Feel free to suggest an alternative!

Reviewer 2: I feel like consumption is a bit more oriented toward usage-related scenarios. In some other places, we’ve used wording like:

> *…associated with a resource or service used or purchased...*
> 
> *…associated with a used or purchased resource or service...*

Here is my attempt to include purchase scenarios as well:

```suggestion
Describes the cost and usage incurred through using or purchasing a provider’s [*resources*](#glossary:resource) or [*services*](#glossary:service).
```

Author: This works for me!  Accepting suggestion.
